src/Client.php

```diff
@@ -189,6 +192,32 @@ final public function addSubscriber(Subscriber $subscriber): void
         $this->manager->addSubscriber($subscriber);
     }
 
+    /**
+     * Executes multiple write operations.
+     *
+     * @see ClientBulkWrite::__construct() for supported options
+     * @param string $databaseName Database name
+     * @param array  $options      Additional options
+     * @throws UnsupportedException if options are unsupported on the selected server
+     * @throws InvalidArgumentException for parameter/option parsing errors
+     * @throws DriverRuntimeException for other driver errors (e.g. connection errors)
+     */
+    public function bulkWrite(BulkWriteCommand|BulkWriteCommandBuilder $bulk, array $options = []): ?BulkWriteCommandResult
```

Comment:
When is `null` returned? I think this should be mentioned in the comment.

Reply:
Currently, `executeBulkWriteCommand()` returns `null` when an unacknowledged write concern is used. 

The alternative would be to return BulkWriteCommandResult object where `isAcknowledged()` returns `false` and no other methods can be called. That's what we do in the legacy `executeBulkWrite()` API.

I'm leaning towards a `null` object since it's consistent with the spec's instruction on _not_ populating `partialResult` on BulkWriteCommandException when zero successful writes have occurred. In the legacy API, the result object could just be empty.

Reply:
I'm now wondering if the current design of having `executeBulkWriteCommand()` return a nullable BulkWriteCommandResult is optimal. Interestingly, the [spec](https://github.com/mongodb/specifications/blob/master/source/crud/bulk-write.md) doesn't explicitly suggest that when it presents a signature for `Client::bulkWrite()`, but the section on [modeling unacknowledged results](https://github.com/mongodb/specifications/blob/master/source/crud/bulk-write.md#unacknowledged-results) ends up suggesting this when it links to the [CRUD spec's guidance](https://github.com/mongodb/specifications/blob/master/source/crud/crud.md#write-results) on modeling unacknowledged results. I was also influenced by libmongoc, which leaves the result null in these cases.

The spec is clear about BulkWriteCommandException having an optional value for the partial write result, so I was planning to use nullable BulkWriteCommandResult for its `$partialResult` property (and getter); however, I'm curious if a BulkWriteCommandResult object in an unacknowledged state would also be preferable there.

In [PHPC-2144](https://jira.mongodb.org/browse/PHPC-2144), we changed PHPC 2.0 to remove the nullable return types for fields on WriteResult when the write was unacknowledged. Instead, we just throw if getters are accessed and the result object is empty (i.e. isAcknowledged() returns false). That made PHPC consistent with the PHPLIB API. My memory is a bit fuzzy, but I remember the nullable return values being more of an annoyance for static analysis. And in the general case where users don't use `w:0`, they can probably ignore `isAcknowledged()` and just assume that exceptions won't be thrown when they access the result.

Reply:
In general, I think it would be preferable to return something other than `null` in all cases, as long as users can deduce what happen (e.g. unacknowledged writes). Returning `null` in some cases creates one "special" condition that users need to check for instead of handling a result in all cases that didn't error.

Reply:
The method `Collection::bulkWrite()` always return an instance of `BulkWriteResult`, I would do the same for `Client::bulkWrite()`.

Reply:
> I think it would be preferable to return something other than null in all cases

How do you feel about the verbose result fields (e.g. `BulkWriteCommandResult::getInsertResults()`) returning a nullable Document? If we have the regular getters (e.g. `getInsertedCount()`) throw when `isAcknowledged()` is false, we can do the same with the others in relation to a `hasVerboseResults()` field. That is one approach suggested in the [bulkWrite spec](https://github.com/mongodb/specifications/blob/master/source/crud/bulk-write.md#summary-vs-verbose-results), although it's this is a level removed from the `Client::bulkWrite()` API so it's less important.

I also didn't have qualms about using a nullable BulkWriteCommandResult for `BulkWriteCommandException::getPartialResult()`, which is directly suggested in the spec.

Reply:
`Client::bulkWrite()` always returns a BulkWriteCommandResult now.

I've left `BulkWriteCommandException::getPartialResult()` as `?BulkWriteCommandResult`, since the result is intended to be unset. While `isAcknowledged()` could also be used in that capacity to distinguish an "empty" result, it wouldn't be semantically correct.

I've also left `?Document` for the optional verbose fields.

Reply:
So you can update the return type of this method.
```suggestion
    public function bulkWrite(BulkWriteCommand|BulkWriteCommandBuilder $bulk, array $options = []): BulkWriteCommandResult
```